src/clean.c

```diff
@@ -0,0 +1,397 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include <ctype.h>
+#include <errno.h>
+#include <getopt.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "swupd.h"
+
+/* Default value if --all is not passed. */
+const int DAYS_TO_KEEP_FILES = 3;
+
+static void print_help()
+{
+	fprintf(stderr,
+		"Usage:\n"
+		"	swupd clean [options]\n"
+		"\n"
+		"Remove cached data used for old updates from state directory.\n"
+		"\n"
+		"Options:\n"
+		"   --all                   Remove all the cached content not only old\n"
+		"   --dry-run               Just print files that would be removed\n"
+		"   -S, --statedir DIR      Specify alternate swupd state directory\n"
+		"   -h, --help              Display this help message\n"
+		"\n");
+}
+
+static struct {
+	int all;
+	int dry_run;
+} options;
+
+static struct {
+	int files_removed;
+} stats;
+
+static struct timespec now;
+
+static const struct option prog_opts[] = {
+	{ "help", no_argument, 0, 'h' },
+	{ "all", no_argument, &options.all, 1 },
+	{ "statedir", required_argument, 0, 'S' },
+	{ "dry-run", no_argument, &options.dry_run, 1 },
+	{ 0, 0, 0, 0 }
+};
+
+static bool parse_options(int argc, char **argv)
+{
+	int opt;
+	while ((opt = getopt_long(argc, argv, "hS:", prog_opts, NULL)) != -1) {
+		switch (opt) {
+		case '?':
+			return false;
+		case 'h':
+			print_help();
+			exit(0);
+		case 'S':
+			if (!optarg || !set_state_dir(optarg)) {
+				fprintf(stderr, "Invalid --statedir argument\n\n");
+				return false;
+			}
+			break;
+		case 0:
+			/* Handle options that don't have shortcut. */
+			break;
+		default:
+			fprintf(stderr, "Error: unrecognized option: -%c,\n\n", opt);
+			return false;
+		}
+	}
+	return true;
+}
+
+typedef bool(remove_predicate_func)(const char *dir, const struct dirent *entry);
+
+/* Remove files from path for which pred returns true.
+ * Currently it doesn't recursively remove directories.
+ *
+ * Unless --all is set, the files are also filtered by their age.
+ */
+static int remove_if(const char *path, remove_predicate_func pred)
```

Comment:
I very much like this functional approach.

I think you can take it further if you implement remove_predicate_func as a closure instead, probably not worth it for the simple cases here but you could create AND and OR functions to combine predicates....

Reply:
Yeah, these approach could be extended if we need more complex cases.

I am assuming by closure you meant passing a data pointer to both remove_if (and then to predicate calls).

Reply:
The usual way I implement a closure in C is to make what you currently have as a function pointer to be a struct pointer, and inside the struct have the function pointer and a void* pointer.

You can then have an AND closure which uses the void * pointer to point to a vector of closures...